src/main/java/net/minecraftforge/common/util/ITeleporter.java

```diff
@@ -19,49 +19,80 @@
  * in a safe location, or generate a return portal, for instance.
- *
+ * <p>
  * See the {@link net.minecraft.world.Teleporter} class, which has
  * been patched to implement this interface, for a vanilla example.
  */
-public interface ITeleporter {
-
+public interface ITeleporter
+{
     /**
      * Called to handle placing the entity in the new world.
-     *
+     * <p>
      * The initial position of the entity will be its
      * position in the origin world, multiplied horizontally
      * by the computed cross-dimensional movement factor.
-     *
+     * <p>
      * Note that the supplied entity has not yet been spawned
      * in the destination world at the time.
      *
-     * @param entity           the entity to be placed
-     * @param currentWorld     the entity's origin
-     * @param destWorld        the entity's destination
-     * @param yaw              the suggested yaw value to apply
+     * @param entity the entity to be placed
+     * @param currentWorld the entity's origin
+     * @param destWorld the entity's destination
+     * @param yaw the suggested yaw value to apply
      * @param repositionEntity a function to reposition the entity, which returns the new entity in the new dimension. This is the vanilla implementation of the dimension travel logic. If the supplied boolean is true, it is attempted to spawn a new portal.
+     *
      * @return the entity in the new World. Vanilla creates for most {@link Entity}s a new instance and copy the data. But <b>you are not allowed</b> to create a new instance for {@link PlayerEntity}s! Move the player and update its state, see {@link ServerPlayerEntity#changeDimension(net.minecraft.world.server.ServerWorld, ITeleporter)}
      */
-    default Entity placeEntity(Entity entity, ServerWorld currentWorld, ServerWorld destWorld, float yaw, Function<Boolean, Entity> repositionEntity) {
-       return repositionEntity.apply(true);
+    default Entity placeEntity(Entity entity, ServerWorld currentWorld, ServerWorld destWorld, float yaw,
+            Function<Boolean, Entity> repositionEntity)
+    {
+        return repositionEntity.apply(true);
     }
-    
-    //used internally to handle vanilla hardcoding
+
+    /**
+     * Is this teleporter the vanilla instance.
+     */
     default boolean isVanilla()
     {
-        return getClass() == Teleporter.class;
+        return false;
+    }
+
+    /**
+     * Finds a portal to teleport to and creates a {@link TeleportationRepositioner.Result} for it.
+     * Return an empty {@link Optional} if no portal was found.
+     */
+    Optional<TeleportationRepositioner.Result> findPortal(ServerWorld fromWorld, ServerWorld toWorld, Entity entity);
+
+    /**
+     * Creates a portal if one doesn't exist and returns the {@link TeleportationRepositioner.Result Result}.
+     */
+    Optional<TeleportationRepositioner.Result> createAndGetPortal(ServerWorld fromWorld, ServerWorld toWorld,
```

Comment:
This is a breaking change as any mods currently using ITeleporter will not have this method implemented if it is not defaulted.